index.html

```diff
@@ -21,8 +21,21 @@ <h1>Hello</h1>
         <p><em><small id="deviceName"></small></em></p>
         <script>
         // handle past and future state updates
+        var El = function (tag, text) {
+            var el = document.createElement(tag);
+            el.innerText = text || '';
+            return el;
+        };
+
         window.webxdc.setUpdateListener(function (update) {
-            document.getElementById('output').innerHTML += "<strong>" + update.payload.name + ":</strong> " + update.payload.msg + "<br>";
+            var output = document.getElementById('output');
+            [
+                El('strong', update.payload.name + ':'),
+                El('span', update.payload.msg),
+                El('br'),
+            ].forEach(function (item) {
+                output.appendChild(item);
+            });
```

Comment:
a comment why innerHTML is better not used would be good here.

Reply:
it would seem to me that the only source that could poison the payload to contain "evil js" is the app itself -- unless you have a chat participant who has a hacked up version of Delta Chat and tries to send some html injection along with app updates. But if an adversary is in a chat group and they send apps, they can anyway do arbitrary things in the webview and also send information back to the attacker.

Reply:
i am a little doubtful to worry about "cheating" chat participants -- it kinds of is a paranoid model that makes things maybe unncessarily harder?

Reply:
In this particular app I don't think there's any information that's worth leaking, and webxdc makes it such that it can only be leaked to the chat context in which the app is embedded.

At worst the chat participant can be annoying and use the XSS to cause other participants with active webviews to reflect the malicious payload back, making for a distributed [fork bomb](https://en.wikipedia.org/wiki/Fork_bomb) that ties up everyone's CPU.

I tested with a basic example of this with `<img src="x" onerror="(function(){ setInterval(function(){ window.webxdc.sendUpdate({ payload: { name: window.webxdc.selfName, msg: 'pewpew' }, info: 'PEWPEW' }, 'PEWPEW'); }, 1); }());" />`

I figure people can always choose to implement their apps however they want, so this isn't making it harder for anyone, just setting a safer example.

Reply:
another approach which I think might work is to define a content-security-policy in a meta tag, like

> `<meta http-equiv="Content-Security-Policy" content="script-src 'self'">`

...which would block any inline code and require that the _hello_ app logic be defined in an external js file.

Reply:
if we should inject that strict content policy by default (using the csp header), we should have a key in the manifest to disable it again, some apps/frameworks might depend on it?

Reply:
Yet another (simpler) option is to escape the HTML payload anywhere injection would be possible:

```diff
diff --git a/index.html b/index.html
index 5edfdc5..6a21e99 100644
--- a/index.html
+++ b/index.html
@@ -20,9 +20,13 @@
         <p id="output"></p>
         <p><em><small id="deviceName"></small></em></p>
         <script>
+        var escapeHTML = function (str) {
+            return (str || "").replace(/</g, '&lt;');
+        };
+
         // handle past and future state updates
         window.webxdc.setUpdateListener(function (update) {
-            document.getElementById('output').innerHTML += "<strong>" + update.payload.name + ":</strong> " + update.payload.msg + "<br>";
+            document.getElementById('output').innerHTML += "<strong>" + escapeHTML(update.payload.name) + ":</strong> " + escapeHTML(update.payload.msg) + "<br>";
         });
 
         function sendMsg() {
@@ -41,7 +45,7 @@
         }
 
         (function () {
-            window.deviceName.innerHTML = 'this is ' + window.webxdc.selfName;
+            window.deviceName.innerHTML = 'this is ' + escapeHTML(window.webxdc.selfName);
         })()
         </script>
     </body>
 ```

Reply:
the latter is much easier to understand for ppl only casually into javascript. or for ppl not into "modern" javascript.

all in all, i agree, escaping should be done somehow to not support coding that is harmful under other, non-webxdc circumstances. in case of webxdc, i so not see real xss issues aport from breaking layout, so, i would not go for some "global" switch that again makes things harder to use and needs ppl more to understand.

Reply:
we already disallow webrequests and one mayor part todays usage of javasript is loading javascript from public CDNs like google/fastly/Akamai/cloudflare and others.
So I don't think it's a bad idea nudging people into the right direction of not using unsafe javascript.
We could describe why we disabled it in trouble shooting (in the docs) and would make unsafe stuff opt-in.
also browsers show errors in the js console if stuff was disallowed by the CSP